### ISSUE TITLE
Memory leak when installing event handlers

### ISSUE THREAD

**Issue by freakboy3742** [pythonnet/pythonnet#1434]
### Environment

-   Pythonnet version: 2.5.2
-   Python version: 3.8.3
-   Operating System: Windows 10 Home (Build 19041.804)
-   .NET Runtime: 4.8.04084 (I think?)

### Details

-   Describe what you were trying to get done.

Develop a native WinForms application.

-   What commands did you run to trigger this issue? 

Any WinForms app that has an event handler should exhibit the problem. Here's a minimal example derived from the HelloApp example in the pythonnet codebase.

```python
import clr

SWF = clr.AddReference("System.Windows.Forms")
print (SWF.Location)
import System.Windows.Forms as WinForms
from System.Drawing import Size, Point


class HelloApp(WinForms.Form):
    def __init__(self):
        self.ClientSize = Size(400, 200)

        self.button = WinForms.Button()
        self.button.Location = Point(50, 50)
        self.button.Size = Size(300, 20)
        self.button.Text = "Click Me!"

        self.button.Click += self.button_Click

        self.Controls.Add(self.button)

    def button_Click(self, sender, args):
        print ("Click")
    
    def run(self):
        WinForms.Application.Run(self)

def main():
    form = HelloApp()
    app = WinForms.Application
    app.Run(form)


if __name__ == '__main__':
    main()
```

Run the app, then really mash the "click me" button (lots - like, dozens of times). Every time you click, the app's memory allocation increases a little bit; I've never seen this allocation be freed by a garbage collection cycle, even if you add a `gc.collect()` call.

This problem was observed in practice with code that [integrates the Python Windows Proactor with the Winforms event loop](https://github.com/beeware/toga/blob/master/src/winforms/toga_winforms/libs/proactor.py). In that code, we're creating and scheduling a `System.Action()` every 5ms to schedule an iteration of the asyncio event loop - and every time it does, there's a small leak. Invoke that 200 times a second, and the leak adds up quickly. However, even simple button presses cause a leak.

-   If there was a crash, please include the traceback here.

No crash - the app works fine; it just (slowly) leaks memory.

**Comment by filmor** [comment]
Could you test your code against the current master? I vaguely remember that fixes for this or something similar were merged a while ago.

**Comment by freakboy3742** [comment]
@filmor Not easily... I don't have a C development environment configured on Windows; I rely on published wheels for all my Windows development. 

Is there any way to get access to a recent dev/prerelease wheel? I can't see any reference to that as an option in the docs...

**Comment by filmor** [comment]
The current master doesn't require C tooling anymore. You only need the .NET SDK. I'll work on making weekly development wheels available as well this week.

**Comment by freakboy3742** [comment]
@filmor I've just tested the current master (23527d11) - I'm not seeing any change in behavior.

**Comment by saroad2** [comment]
Hey, is there any development with this issue?
We have severe memory leaks in [Eddington-GUI](https://github.com/EddLabs/eddington-gui) when trying to convert to newest version of Pythonnet.

**Comment by saroad2** [comment]
I don't want to nag, but any update would be much appreciated.
If there's something I can do, maybe a hint where the bug could be in, I would love to hear and maybe give it a go trying to fix this.
This bug really halted all progress in Eddington, and I would love to see it fixed.

**Comment by freakboy3742** [comment]
I'll add that if this is a situation where money will accelerate progress, I may be in a position to help out.

**Comment by lostmsu** [comment]
@freakboy3742 I just run a short test locally on `master`: not seeing any memory leaks when an event handler is invoked in a loop. .NET has its own GC, which is called randomly unless you force it with `System.GC.Collect()` (and maybe also `System.GC.WaitForPendingFinalizers()` to disown Python objects), which would explain why the mem usage slowly grows.

@saroad2 please, narrow down your repro steps if you think there's a problem with Python.NET

**Comment by freakboy3742** [comment]
@lostmsu I've just tried master (dc3c8188), with .NET SDK 5.0.400, and I'm seeing the same behavior with the test program I originally provided. Using that test program, mashing the button, I see a slow leak (as observed with the Windows Resource Monitor). It's difficult to quantify exactly how big the leak is on a single event, but it's on the order of 4-15kB per click.

Using the test program from the original report, I observe a starting Commit memory usage of 40,884kB; if I press the button 100 times, I see Commit memory usage of 42,084kB.

I've also tried adding a second button to that test program that invokes `gc.collect()`, `System.GC.Collect()`, and `System.GC.WaitForPendingFinalizers()`. Pressing that button has no appreciable drop in Commit memory. Sometimes, I even see an *increase* when pressing this button - but this behavior is not consistent.

I will note that it the behavior of master seems a *lot* better than the behavior of 2.5.2.

**Comment by saroad2** [comment]
I can confirm that the problem is still happening using @freakboy3742 example.
Here is what you can do to see it:

1. Install memory profiler and matplotlib using `pip install memory_profiler matplotlib`
2. Write the code example into a file named `demo.py`
3. Run `mprof run --include-children demo.py`
4. Click The button multiple times
5. Exit the program
6. Run `mprof plot` and see the memory leak

![Demo Leak](https://user-images.githubusercontent.com/19425795/132974404-71becc8b-18d4-472f-9c4c-fc2547103866.png)

**Comment by lostmsu** [comment]
@freakboy3742 I do not expect GC to release memory to the operating system, because they don't do that all the time. However, if you press the button say 10 times, then call GC, the next 10 times you press the button there's likely going to be no increase in memory usage.

**Comment by freakboy3742** [comment]
@lostmsu Thanks for the clarification. I've made a change to my "real world" case that invokes GC (on both Python and .NET) every minute, and that seems to have addressed the problem (using current master/dc3c818). There's still lots of fluctuation in memory usage (+/- 2MB); but there's no increase over the long term. I'm still seeing the leak in 2.5.2, but I can appreciate that your development focus is on the 3.0 branch.

Are you able to elaborate on why this manual GC handling is required? In other contexts, I'd generally consider manually invoking GC to be a bit of a code smell - it may work, but the fact that it is needed at all is an indicator you've got something bigger going wrong. Is this not the case with .NET? Is there something that I could be doing (other than manually invoking GC) to convince .NET to clean up after itself? More specifically - why does the sample program provided as part of the original report *not* GC itself cleanly without manual assistance?

**Comment by lostmsu** [comment]
@freakboy3742 manual GC handling is not required. GC will run when certain conditions are met. Roughly: every X bytes allocated, or when system-wide free RAM is low. 20Kb is a very low threshold.
